src/services/refactors/moveToNewFile.ts

```diff
@@ -0,0 +1,596 @@
+
+                const top = getTopLevelDeclarationStatement(decl);
+                if (markSeenTop(top)) {
+                    addExportToChanges(oldFile, top, changes, useEs6ModuleSyntax);
+                }
+                if (hasModifier(decl, ModifierFlags.Default)) {
+                    oldFileDefault = name;
+                }
+                else {
+                    oldFileNamedImports.push(name.text);
+                }
+            }
+        });
+
+        append(copiedOldImports, makeImportOrRequire(oldFileDefault, oldFileNamedImports, removeFileExtension(getBaseFileName(oldFile.fileName)), useEs6ModuleSyntax));
+        return copiedOldImports;
+    }
+
+    function makeUniqueModuleName(moduleName: string, extension: string, inDirectory: string, host: LanguageServiceHost): string {
+        let newModuleName = moduleName;
+        for (let i = 1; ; i++) {
+            const name = combinePaths(inDirectory, newModuleName + extension);
+            if (!host.fileExists(name)) return newModuleName;
+            newModuleName = `${moduleName}.${i}`;
+        }
+    }
+
+    function getNewModuleName(movedSymbols: ReadonlySymbolSet): string {
+        return movedSymbols.forEachEntry(symbolNameNoDefault) || "newFile";
+    }
+
+    interface UsageInfo {
+        // Symbols whose declarations are moved from the old file to the new file.
+        readonly movedSymbols: ReadonlySymbolSet;
+
+        // Symbols declared in the old file that must be imported by the new file. (May not already be exported.)
+        readonly newFileImportsFromOldFile: ReadonlySymbolSet;
+        // Subset of movedSymbols that are still used elsewhere in the old file and must be imported back.
+        readonly oldFileImportsFromNewFile: ReadonlySymbolSet;
+
+        readonly oldImportsNeededByNewFile: ReadonlySymbolSet;
+        // Subset of oldImportsNeededByNewFile that are will no longer be used in the old file.
+        readonly unusedImportsFromOldFile: ReadonlySymbolSet;
+    }
+    function getUsageInfo(oldFile: SourceFile, toMove: ReadonlyArray<Statement>, checker: TypeChecker): UsageInfo {
+        const movedSymbols = new SymbolSet();
+        const oldImportsNeededByNewFile = new SymbolSet();
+        const newFileImportsFromOldFile = new SymbolSet();
+
+        for (const statement of toMove) {
+            forEachTopLevelDeclaration(statement, decl => {
+                movedSymbols.add(Debug.assertDefined(isExpressionStatement(decl) ? checker.getSymbolAtLocation(decl.expression.left) : decl.symbol));
+            });
+
+            forEachReference(statement, checker, symbol => {
+                if (!symbol.declarations) return;
+                for (const decl of symbol.declarations) {
+                    if (isInImport(decl)) {
+                        oldImportsNeededByNewFile.add(symbol);
+                    }
+                    else if (isTopLevelDeclaration(decl) && !movedSymbols.has(symbol)) {

```

Comment:
what if it is not in movedSymbols just case we have not visited it yet, e.g. a forward reference to a type or an interface.. or even a class, that exists later on in the selected range?